### PR TITLE
[29698] Playback History

### PR DESCRIPTION
**Pull request by akrol95**

### CODE REVIEW

File: BMM.Core/Constants/LoggingEvents.cs [hunk 1]
```diff
@@ -0,0 +1,7 @@
+namespace BMM.Core.Constants
+{
+    public class LoggingEvents
+    {
+        public const string ViewModelOpenedFormat = "{0} opened";
```
Comment (kkuepper) [inline BMM.Core/Constants/LoggingEvents.cs:5]: There's already a similar class: BMM.Core.Implementations.Analytics.Event
I think I like your namespace better. Nevertheless they should be merged.

File: BMM.Core/Implementations/Persistence/Interfaces/IBlobCacheWrapper.cs [hunk 2]
```diff
@@ -0,0 +1,10 @@
+using System.Threading.Tasks;
+
+namespace BMM.Core.Implementations.Persistence.Interfaces
+{
+    public interface IBlobCacheWrapper
```
Comment (kkuepper) [inline BMM.Core/Implementations/Persistence/Interfaces/IBlobCacheWrapper.cs:5]: Isn't it better to just call this ICache and BlobCacheWrapper is just the implementation which decides to use BlobCache to persist the cache.

File: BMM.Core/Implementations/PlayObserver/Storage/TrackPlayedStorage.cs [hunk 3]
```diff
@@ -11,7 +11,7 @@ namespace BMM.Core.Implementations.PlayObserver.Storage
 {
     public class TrackPlayedStorage : ITrackPlayedStorage
     {
-        private const string StorageKey = StorageKeys.FinishedTrackPlayedEvents;
+        private readonly string _storageKey = StorageKeys.FinishedTrackPlayedEvents;
```
Comment (kkuepper) [inline BMM.Core/Implementations/PlayObserver/Storage/TrackPlayedStorage.cs:14]: why did you change this?
Can you at least make it static so that we can keep the PascalCase naming and don't need the underscore?

File: BMM.Core/Implementations/Player/PlaybackHistoryService.cs [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+using BMM.Core.Implementations.Player.Interfaces;
+using BMM.Core.Models.PlaybackHistory;
+using BMM.Core.Models.PlaybackHistory.Interfaces;
+
+namespace BMM.Core.Implementations.Player
+{
+    public class PlaybackHistoryService : IPlaybackHistoryService
+    {
+        public const int MaxEntries = 100;
+        private readonly IBlobCacheWrapper _blobCacheWrapper;
```
Comment (kkuepper) [inline BMM.Core/Implementations/Player/PlaybackHistoryService.cs:20]: ICache _cache is better. See other comment

File: BMM.Core/Implementations/Player/PlaybackHistoryService.cs [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+        public PlaybackHistoryService(
+            IBlobCacheWrapper blobCacheWrapper)
+        {
+            _blobCacheWrapper = blobCacheWrapper;
+        }
+
+        public async void AddPlayedTrack(IMediaTrack mediaTrack)
```
Comment (kkuepper) [inline BMM.Core/Implementations/Player/PlaybackHistoryService.cs:28]: Isn't async void dangerous and should be prevented if possible?
Also, do we need to worry about concurrency? e.g. quickly clicking multiple tracks.

File: BMM.Core/Implementations/Player/PlaybackHistoryService.cs [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+                .LastOrDefault()
+                ?.MediaTrack;
+
+            if (lastTrack != null && lastTrack.Id == mediaTrack.Id)
+                return;
+
+            if (playbackHistory.Count >= MaxEntries)
+                playbackHistory.RemoveAt(0);
+
+            playbackHistory.Add(new PlaybackHistoryEntry((Track) mediaTrack, DateTime.UtcNow));
```
Comment (kkuepper) [inline BMM.Core/Implementations/Player/PlaybackHistoryService.cs:42]: The TrackPlayedEvent already creates a timestamp. Can we reuse that one instead of creating a new one?

File: BMM.Core/Implementations/Player/PlaybackHistoryService.cs [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+        public async Task<IEnumerable<IPlaybackHistoryGroup>> GetAll()
+        {
+            var playbackHistory = await GetPlaybackHistory();
+
+            if (!playbackHistory.Any())
+                return Enumerable.Empty<IPlaybackHistoryGroup>();
+
+            var result = playbackHistory
+                .OrderByDescending(x => x.PlayedAtUTC)
+                .GroupBy(l => new
```
Comment (kkuepper) [inline BMM.Core/Implementations/Player/PlaybackHistoryService.cs:58]: Hm. We decided to group by day because we have a simple way of showing time information with just introducing a header. But I can easily think of other ways of presenting this information that would require a different grouping.
What I want to say is that the grouping is closely related to the UI and I think it should not bleed into this class. This class should be more independent of how we want to present the information.
I would suggest moving it into the guarded action

File: BMM.Core/Implementations/Player/ViewModelHandlingMediaPlayerDecorator.cs [hunk 8]
```diff
@@ -87,13 +91,15 @@ public async Task Play(IList<IMediaTrack> mediaFiles, IMediaTrack currentTrack,
         {
             if (!ShowErrorIfOutdatedLiveRadio(currentTrack))
             {
+                _playbackHistoryService.AddPlayedTrack(currentTrack);
```
Comment (kkuepper) [inline BMM.Core/Implementations/Player/ViewModelHandlingMediaPlayerDecorator.cs:94]: what's the performance of this action? accessing the blobcache easily takes 100ms because of Json deserialization. But playing the song is a very time-sensitive action.

Comment (kkuepper) [inline BMM.Core/Implementations/Player/ViewModelHandlingMediaPlayerDecorator.cs:94]: Sorry, I probably didn't explain this properly.
The PBI says that we should store "latest position". My idea was that the user can continue the track where he left off. That's very relevant for speeches that are often around 30 minutes long.
However to store the latest position we would need to add the track to the history AFTER he is done playing.

We also need something in the ViewModel to continue playing where the user left off.

File: BMM.Core/Models/PlaybackHistory/Interfaces/IPlaybackHistoryEntry.cs [hunk 9]
```diff
@@ -0,0 +1,11 @@
+using System;
+using BMM.Api.Implementation.Models;
+
+namespace BMM.Core.Models.PlaybackHistory.Interfaces
+{
+    public interface IPlaybackHistoryEntry
+    {
+        Track MediaTrack { get; }
+        DateTime PlayedAtUTC { get; }
```
Comment (kkuepper) [inline BMM.Core/Models/PlaybackHistory/Interfaces/IPlaybackHistoryEntry.cs:9]: Like the name and putting utc in the name

File: BMM.Core/ViewModels/ExploreNewestViewModel.cs [hunk 10]
```diff
@@ -37,18 +41,22 @@ public class ExploreNewestViewModel : DocumentsViewModel
...
             _bmmLanguageBinder = bmmLanguageBinder;
+            _navigateToViewModelAction = navigateToViewModelAction;
             FraKaareTeaserViewModel = Mvx.IoCProvider.IoCConstruct<FraKaareTeaserViewModel>();
             AslaksenTeaserViewModel = Mvx.IoCProvider.IoCConstruct<AslaksenTeaserViewModel>();
             RadioViewModel = Mvx.IoCProvider.IoCConstruct<ExploreRadioViewModel>();
             _listeningStreakToken = messenger.Subscribe<ListeningStreakChangedMessage>(ListeningStreakChanged);
             TrackInfoProvider = new TypeKnownTrackInfoProvider();
         }
 
+        public IMvxAsyncCommand<Type> NavigateToViewModelCommand => _navigateToViewModelAction.Command;
```
Comment (kkuepper) [inline BMM.Core/ViewModels/ExploreNewestViewModel.cs:58]: I don't get it. Why is this needed?

File: BMM.Core/ViewModels/PlaybackHistoryViewModel.cs [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+            var mediaTracksList = list
+                .OfType<IMediaTrack>()
+                .ToList();
+
+            await _mediaPlayer.Play(mediaTracksList, mediaTrack, GetType().Name);
+        }
+
+        public override async Task<IEnumerable<Document>> LoadItems(CachePolicy policy = CachePolicy.UseCacheAndRefreshOutdated)
+        {
+            return await _preparePlaybackHistoryAction.ExecureGuarded();
```
Comment (kkuepper) [inline BMM.Core/ViewModels/PlaybackHistoryViewModel.cs:53]: Typo "execure"

**Review by kkuepper — CHANGES_REQUESTED** [review]

File: BMM.Core/Implementations/Persistence/Interfaces/IBlobCacheWrapper.cs (hunk 2 above)
Comment (kkuepper) [reply BMM.Core/Implementations/Persistence/Interfaces/IBlobCacheWrapper.cs:5]: please check how many methods are actually used

File: BMM.Core/Implementations/PlayObserver/Storage/TrackPlayedStorage.cs (hunk 3 above)
Comment (kkuepper) [reply BMM.Core/Implementations/PlayObserver/Storage/TrackPlayedStorage.cs:14]: doesn't really matter

File: BMM.Core/Implementations/Player/PlaybackHistoryService.cs (hunk 5 above)
Comment (akrol95) [reply BMM.Core/Implementations/Player/PlaybackHistoryService.cs]: Added semaphore

File: BMM.Core/Implementations/Player/ViewModelHandlingMediaPlayerDecorator.cs (hunk 8 above)
Comment (akrol95) [reply BMM.Core/Implementations/Player/ViewModelHandlingMediaPlayerDecorator.cs]: It depends on device, I tested on my samsung and it takes less than 100ms with maximum 100 entries added. I haven't noticed any negative impact in UI.

File: BMM.Api/Implementation/Models/Track.cs [hunk 12]
```diff
@@ -52,16 +52,44 @@ public Track()
         public string InternalTitle { get; set; }
 
         [JsonIgnore]
-        public string GetUniqueKey => Id + "_" + Language;
+        public string GetUniqueKey => $"{Id}_{Language}";
+
+        public long LastPosition { get; set; }
+
+        public DateTime LastPlayedAtUTC { get; set; } = DateTime.MinValue;
```
Comment (kkuepper) [inline BMM.Api/Implementation/Models/Track.cs:59]: isn't it better to make this nullable?

File: BMM.Api/Implementation/Models/Track.cs [hunk 13]
```diff
@@ -52,16 +52,44 @@ public Track()
         public string InternalTitle { get; set; }
 
         [JsonIgnore]
-        public string GetUniqueKey => Id + "_" + Language;
+        public string GetUniqueKey => $"{Id}_{Language}";
+
+        public long LastPosition { get; set; }
```
Comment (kkuepper) [inline BMM.Api/Implementation/Models/Track.cs:57]: Nullable long?

File: BMM.Core/Extensions/ObjectExtensions.cs [hunk 14]
```diff
@@ -0,0 +1,16 @@
+using Newtonsoft.Json;
+
+namespace BMM.Core.Extensions
+{
+    public static class ObjectExtensions
+    {
+        /// <summary>
+        /// Simple Copy method using serializing and deserializing JSON.
+        /// </summary>
+        public static T CopyBySerialization<T>(this T objectToCopy)
```
Comment (kkuepper) [inline BMM.Core/Extensions/ObjectExtensions.cs:10]: are you using this a lot? especially on older android phones this might be expensive

File: BMM.Core/Implementations/PlayObserver/PersistedEventWriter.cs [hunk 15]
```diff
@@ -21,35 +23,44 @@ public class PersistedEventWriter : IDelayedStartupTask
...
             _logger.Info("PersistedEventLogger", "Logging Track played after startup");
             _analytics.LogEvent("Log Track played after startup",
                 new Dictionary<string, object>
                 {
-                    {"TrackId", storedEvent.TrackId},
-                    {"TimestampStart", storedEvent.TimestampStart},
-                    {"TimestampEnd", storedEvent.TimestampEnd},
-                    {"SpentTime", storedEvent.SpentTime},
-                    {"UniqueSecondsListened", storedEvent.UniqueSecondsListened}
+                    {nameof(storedEvent.TrackId), storedEvent.TrackId},
```
Comment (kkuepper) [inline BMM.Core/Implementations/PlayObserver/PersistedEventWriter.cs:55]: I thought about using nameof in this case before. But we have several analytics charts that rely on these names. Therefore I decided not to do it, so that if we rename the property, we still get the correct string to the backend without breaking those charts.

File: BMM.Core/Implementations/PlayObserver/PlayStatistics.cs [hunk 16]
```diff
@@ -181,8 +193,9 @@ private async Task LogPlayedTrack()
                 LogListenedPortionsIfUniqueSecondsListenedAreGreaterThanSpentTime(measurements.UniqueSecondsListened, measurements.SpentTime);
 
                 var ev = ComposeEvent(measurements);
-                await WriteEvent(ev);
+                await _playbackHistoryService.AddPlayedTrack((Track)CurrentTrack, ev.LastPosition, ev.TimestampStart ?? DateTime.UtcNow);
```
Comment (kkuepper) [inline BMM.Core/Implementations/PlayObserver/PlayStatistics.cs:196]: how can TimestampStart be null?

File: BMM.Core/Translation/en/main.json [hunk 17]
```diff
@@ -436,6 +438,10 @@
   "CopyrightViewModel": {
     "Title": "Copyright"
   },
+  "PlaybackHistoryViewModel": {
+    "Title": "Playback History",
+    "NoHistoryYet": "No playback history yet."
```
Comment (kkuepper) [inline BMM.Core/Translation/en/main.json:443]: +1 for thinking of that use case

File: BMM.Core/Translation/en/main.json [hunk 18]
```diff
@@ -436,6 +438,10 @@
   "CopyrightViewModel": {
     "Title": "Copyright"
   },
+  "PlaybackHistoryViewModel": {
+    "Title": "Playback History",
```
Comment (kkuepper) [inline BMM.Core/Translation/en/main.json:442]: Can you change that to "Recently played" please?

File: BMM.Api/Implementation/Models/Track.cs (hunk 13 above)
Comment (akrol95) [reply BMM.Api/Implementation/Models/Track.cs:57]: Default value for long is 0, so it fits the needs, because if LastPosition is not set we start from 0, so I would leave it as it is.

File: BMM.Core/Extensions/ObjectExtensions.cs (hunk 14 above)
Comment (akrol95) [reply BMM.Core/Extensions/ObjectExtensions.cs:10]: It is used to copy single Track object before saving it to persistent memory. As the operation does not touch I/O (everything is in memory), and operates on pretty small single object i don't think it's big overhead.

File: BMM.Core/Implementations/PlayObserver/PersistedEventWriter.cs (hunk 15 above)
Comment (akrol95) [reply BMM.Core/Implementations/PlayObserver/PersistedEventWriter.cs:55]: Ok, reverted

**Review by kkuepper — APPROVED** [review]
